src/main/java/de/tum/cit/aet/artemis/core/service/ScheduleService.java

```diff
@@ -36,34 +53,153 @@ public class ScheduleService {
+        scheduler.setPoolSize(1);
+        scheduler.setVirtualThreads(true);
+        scheduler.initialize();
+        this.taskScheduler = scheduler;
     }
 
-    private void addScheduledTask(Exercise exercise, ExerciseLifecycle lifecycle, Set<ScheduledFuture<?>> futures) {
-        Tuple<Long, ExerciseLifecycle> taskId = new Tuple<>(exercise.getId(), lifecycle);
-        scheduledExerciseTasks.put(taskId, futures);
+    /**
+     * Get all scheduled exercise events.
+     *
+     * @param pageable the pagination information
+     * @return a page of scheduled exercise events
+     */
+    public Page<ScheduledExerciseEvent> findAllExerciseEvents(Pageable pageable) {
+        // Flatten the map into a list of ScheduledExerciseEvent
+        var allEvents = scheduledExerciseTasks.entrySet().stream().flatMap(entry -> entry.getValue().stream().map(task -> {
+            // Calculate the scheduled time from the future's delay
+            var scheduledTime = ZonedDateTime.now().plusSeconds(task.future().getDelay(TimeUnit.SECONDS));
+            return new ScheduledExerciseEvent(entry.getKey().exerciseId(), entry.getKey().lifecycle(), task.name(), scheduledTime, task.future().state());
+        })).sorted(Comparator.comparing(ScheduledExerciseEvent::scheduledTime)).toList();
+
+        // Apply pagination
+        int start = (int) pageable.getOffset();
+        int end = Math.min(start + pageable.getPageSize(), allEvents.size());
+        List<ScheduledExerciseEvent> paginatedEvents = start < allEvents.size() ? allEvents.subList(start, end) : new ArrayList<>();
+
+        return new PageImpl<>(paginatedEvents, pageable, allEvents.size());
     }
 
-    private void removeScheduledTask(Long exerciseId, ExerciseLifecycle lifecycle) {
-        Tuple<Long, ExerciseLifecycle> taskId = new Tuple<>(exerciseId, lifecycle);
-        scheduledExerciseTasks.remove(taskId);
+    /**
+     * Initializes and schedules periodic logging and cleanup tasks for scheduled exercises
+     * and participation tasks. This method is triggered automatically when the application
+     * is fully started.
+     *
+     * <p>
+     * Every 15 seconds, this method:
+     * <ul>
+     * <li>Logs the total number of scheduled exercise and participation tasks.</li>
+     * <li>Iterates through scheduled exercise tasks and logs their details, including
+     * execution time and state. It removes tasks that are no longer running.</li>
+     * <li>Iterates through scheduled participation tasks and logs their details,
+     * including execution time and state. It removes tasks that are no longer running.</li>
+     * <li>Cleans up empty entries from both task maps to avoid memory leaks.</li>
+     * </ul>
+     *
+     * <p>
+     * The scheduling mechanism ensures that outdated or completed tasks do not persist in
+     * memory unnecessarily while maintaining visibility into scheduled tasks.
+     */
+    @EventListener(ApplicationReadyEvent.class)
+    public void startup() {
```

Comment:
this method contains some duplications. might be fine given the complexity of the method to keep things simple

Reply:
I reduced code duplication with an interface for the two records